zerver/worker/queue_processors.py

```diff
@@ -857,24 +857,28 @@ class FetchLinksEmbedData(QueueProcessingWorker):
     CONSUME_ITERATIONS_BEFORE_UPDATE_STATS_NUM = 1
 
     def consume(self, event: Mapping[str, Any]) -> None:
-        url_embed_data: Dict[str, Optional[UrlEmbedData]] = {}
-        for url in event["urls"]:
-            start_time = time.time()
-            url_embed_data[url] = url_preview.get_link_embed_data(url)
-            logging.info(
-                "Time spent on get_link_embed_data for %s: %s", url, time.time() - start_time
-            )
+        with transaction.atomic():
+            try:
+                message = Message.objects.select_for_update().get(id=event["message_id"])
+            except Message.DoesNotExist:
+                # Message may have been deleted
+                return
 
-        message = Message.objects.get(id=event["message_id"])
-        # If the message changed, we will run this task after updating the message
-        # in zerver.actions.message_edit.check_update_message
-        if message.content != event["message_content"]:
-            return
-        if message.content is not None:
-            query = UserMessage.objects.filter(
-                message=message.id,
-            )
-            message_user_ids = set(query.values_list("user_profile_id", flat=True))
+            # If the message changed, we will run this task after updating the message
+            # in zerver.actions.message_edit.check_update_message
+            if message.content != event["message_content"]:
+                return
+
+            if message.content is None:
+                return
+
+            url_embed_data: Dict[str, Optional[UrlEmbedData]] = {}
+            for url in event["urls"]:
+                start_time = time.time()
+                url_embed_data[url] = url_preview.get_link_embed_data(url)
+                logging.info(
+                    "Time spent on get_link_embed_data for %s: %s", url, time.time() - start_time
+                )
```

Comment:
I worry that this ordering results in our holding the lock on the `Message` object for several seconds if these HTTP requests are somewhat slow. I think it might be worth doing this block outside the transaction, even though it's potentially a waste of time, to avoid that downside. (The potential stability risks from holding locks for a long time seem much more significant than the risk of extra work when racing with deleted/edited messages).

Reply:
Yeah, that's a reasonable worry.  We don't do much locking or mutation of message rows at current, but there's no reason to make that worse for us in the future.

Repushed with that pulled outside the block, and more words about that choice in the commit message.